From 4b97531e953b0f6d7ba40780f3e962059473e68b Mon Sep 17 00:00:00 2001
From: Ulf Samuelsson <ulf@emagii.com>
Date: Tue, 7 Feb 2023 19:20:23 +0100
Subject: [PATCH 2/4] ldgram.y: Add ASCII and ASCIZ commands

---
 ld/ldgram.y | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/ld/ldgram.y b/ld/ldgram.y
index 11c2f219..4bbda989 100644
--- a/ld/ldgram.y
+++ b/ld/ldgram.y
@@ -124,7 +124,7 @@ static int error_index;
 %right UNARY
 %token END
 %left <token> '('
-%token <token> ALIGN_K BLOCK BIND QUAD SQUAD LONG SHORT BYTE
+%token <token> ALIGN_K BLOCK BIND QUAD SQUAD LONG SHORT BYTE ASCII ASCIZ
 %token SECTIONS PHDRS INSERT_K AFTER BEFORE
 %token DATA_SEGMENT_ALIGN DATA_SEGMENT_RELRO_END DATA_SEGMENT_END
 %token SORT_BY_NAME SORT_BY_ALIGNMENT SORT_NONE
@@ -667,6 +667,14 @@ statement:
 		  lang_add_data ((int) $1, $3);
 		}
 
+	| ASCII INT ',' NAME
+		{
+		  lang_add_string($2.integer, $4);
+		}
+	| ASCIZ NAME
+		{
+		  lang_add_stringz($2);
+		}
 	| FILL '(' fill_exp ')'
 		{
 		  lang_add_fill ($3);
-- 
2.17.1

